desc/optimize/optimizer.py

```diff
@@ -345,11 +345,15 @@ def _parse_constraints(constraints):
     """
     if not isinstance(constraints, (tuple, list)):
         constraints = (constraints,)
+    # we treat linear bound constraints as nonlinear since they can't be easily
+    # factorized like linear equality constraints
     linear_constraints = tuple(
-        constraint for constraint in constraints if constraint.linear
+        constraint
+        for constraint in constraints
+        if (constraint.linear and (constraint.bounds is None))
     )
     nonlinear_constraints = tuple(
-        constraint for constraint in constraints if not constraint.linear
+        constraint for constraint in constraints if constraint not in linear_constraints
```

Comment:
this doesn't need (constraint.bounds is not None)? I guess that would be redundant?

Reply:
yeah we already get all the non-bounded constraints into `linear_constraints` and this just gets whatever is leftover